### PR TITLE
[BugFix] Fix stats are not correct in skew join rule for partitioned hive table

### PR DESCRIPTION
**Pull request by Youngwb**
## Why I'm doing:
at skew join rule, row count in statistics is not correct when hive table is partitioned table
## What I'm doing:
1. Get all partition statistics at skew join rule (beacuse this rule run before prune partition rule, do not known the selected partitions)
2. If can not get basic statistics from internal statistics(_statistics_ db), get table statistics from connector metadata 
Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/connector/PartitionUtil.java [hunk 1]
```diff
@@ -243,6 +243,25 @@ public static List<String> getFilteredPartitionKeys(List<String> partitionNames,
...
+
+    public static List<PartitionKey> getPartitionKeys(Table table) {
+        List<PartitionKey> partitionKeys = Lists.newArrayList();
+        if (table.isUnPartitioned()) {
+            partitionKeys.add(new PartitionKey());
+        } else {
+            try {
+                for (String partitionName : getPartitionNames(table)) {
+                    partitionKeys.add(
+                            createPartitionKey(toPartitionValues(partitionName), getPartitionColumns(table), table));
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/connector/PartitionUtil.java:255]: The `getPartitionColumns` can be moved out of the for loop ?

**Review by packy92 — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/connector/PartitionUtil.java (hunk 1 above)
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/connector/PartitionUtil.java]: Done

**Review by murphyatwork — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java [hunk 2]
```diff
@@ -581,7 +582,10 @@ public Void computeHMSTableScanNode(Operator node, ExpressionContext context, Ta
             } else {
                 predicates = ((PhysicalScanOperator) node).getScanOperatorPredicates();
             }
-            List<PartitionKey> partitionKeys = predicates.getSelectedPartitionKeys();
+            // If partition pruned, we should use the selected partition keys to estimate the statistics,
+            // otherwise, we should use all partition keys to estimate the statistics.
+            List<PartitionKey> partitionKeys = predicates.hasPrunedPartition() ? predicates.getSelectedPartitionKeys() :
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java:587]: should we get all partition names? There should be a performance issue here

File: fe/fe-core/src/main/java/com/starrocks/server/MetadataMgr.java [hunk 3]
```diff
@@ -688,22 +689,44 @@ public Statistics getTableStatistics(OptimizerContext session,
...
                         session, table, columns, partitionKeys, predicate, limit)).orElse(null);
+                if (connectorBasicStats != null && internalStatistics != null &&
+                        internalStatistics.getColumnStatistics().values().stream().anyMatch(
+                                columnStatistic -> columnStatistic.getHistogram() != null)) {
+                    // combine connector metadata basic stats and histogram from internal stats
+                    Map<ColumnRefOperator, ColumnStatistic> internalColumnStatsMap = internalStatistics.getColumnStatistics();
+                    Map<ColumnRefOperator, ColumnStatistic> combinedColumnStatsMap = Maps.newHashMap();
+                    connectorBasicStats.getColumnStatistics().forEach((key, value) -> {
+                        if (internalColumnStatsMap.containsKey(key)) {
+                            combinedColumnStatsMap.put(key, ColumnStatistic.buildFrom(value).
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/server/MetadataMgr.java:714]: why not add others internal stats to combinedColumnStats?

Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/server/MetadataMgr.java:714]: because  internal stats is unknown here

**Review by stephen-shelby — APPROVED** [review]

**Pull request merged**